src/Features/CSharp/Portable/Copilot/CSharpCopilotNotImplementedMethodFixProvider.Analyzer.cs

```diff
@@ -0,0 +1,157 @@
+
+internal sealed partial class CSharpCopilotNotImplementedMethodFixProvider
+{
+    private static class DocumentAnalyzer
+    {
+        private const int MaxMethodLength = 1024;
+        private const int ContextLineCount = 10;
+
+        public static async Task<MethodImplementationProposal?> AnalyzeDocumentAsync(Document document, SyntaxNode throwNode, CancellationToken cancellationToken)
+        {
+            // Get the semantic model and syntax root
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // Find the containing method declaration
+            var methodDeclaration = throwNode.Parent.FirstAncestorOrSelf<MethodDeclarationSyntax>();
+
+            // Get symbol information
+            var methodSymbol = semanticModel.GetRequiredDeclaredSymbol(methodDeclaration, cancellationToken);
+
+            // Find references
+            var references = await FindReferencesAsync(methodSymbol, document, cancellationToken).ConfigureAwait(false);
+            var referenceCount = references.Sum(r => r.Locations.Count());
+
+            // Get top 2 surrounding code snippets
+            var topReferences = references
+                .SelectMany(r => r.Locations)
+                .OrderBy(l => l.Location.SourceSpan.Length)
+                .Take(2)
+                .Select(async l =>
+                {
+                    var refDocument = document.Project.Solution.GetDocument(l.Document.Id);
+                    if (refDocument == null)
+                        return null;
+                    var refRoot = await refDocument.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                    if (refRoot == null)
+                        return null;
+
+                    var referenceNode = refRoot.FindNode(l.Location.SourceSpan);
+                    var containingMethod = referenceNode.Ancestors().OfType<MethodDeclarationSyntax>().FirstOrDefault();
+                    var refText = await refDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+                    var contextSpan = GetContextSpan(refText, l.Location.SourceSpan, containingMethod);
+                    return new MethodImplementationReferenceContext
+                    {
+                        FileName = refDocument.Name,
+                        SurroundingCode = refText.ToString(contextSpan)
+                    };
+                })
+                .Where(x => x != null)
+                .ToList();
+
+            var referenceContexts = await Task.WhenAll(topReferences).ConfigureAwait(false);
+
+            // Get previous and next tokens for context
+            var previousToken = methodDeclaration.Modifiers.Count > 0
+                ? methodDeclaration.Modifiers[0]
+                : methodDeclaration.GetFirstToken();
```

Comment:
i don't understand the point of this.  why not always do GetFirsrtToken?  why the Modifiers work?